### PR TITLE
Preparation for introducing the SOF ALSA plugin

### PR DESCRIPTION
**Pull request by ranj063**
The SOF ALSA plugin will be used for running SOF topology pipelines to run on the host for development/debugging purposes.
The changes in this PR include preparatory patches for:
 - Building shared libraries for SOF modules
 - Posix headers for host build
 - Addition of missing platform definitions for library build

### CODE REVIEW

File: posix/include/rtos/alloc.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
Comment (btian1) [inline posix/include/rtos/alloc.h:1]: not able to include those headers? duplicate is not a good option.

Comment (ranj063) [reply posix/include/rtos/alloc.h:1]: @btian1 this is for host build for testbench/tplg parser/plugin etc where zephyr/xtos doesnt feature

Comment (btian1) [reply posix/include/rtos/alloc.h:1]: so you mean plugin build does not have xtos/zephyr included?

File: tools/testbench/common_test.c [hunk 2]
```diff
@@ -264,14 +264,14 @@ int tb_pipeline_params(struct testbench_prm *tp, struct ipc *ipc, struct pipelin
...
-	if (debug)
+	if (host_trace_level >= LOG_LEVEL_DEBUG)
 		printf("debug: %s", message);
 }
 
 /* enable trace in testbench */
 void tb_enable_trace(bool enable)
 {
-	test_bench_trace = enable;
+	host_trace_level = enable;
```
Comment (lyakh) [inline tools/testbench/common_test.c:274]: confused by this commit: (1) where `host_trace_level` is defined (apparently, it is defined somewhere, since this all works, or maybe it doesn't need to be, and (2) doesn't the above mean, that we set `host_trace_level` to either `LOG_LEVEL_ERROR` or to 0, which isn't even a valid error level?

File: src/platform/library/lib/trace.c [hunk 3]
```diff
@@ -12,8 +12,7 @@
 #include <sof/trace/dma-trace.h>
 
 /* enable trace by default in testbench */
-int test_bench_trace = 1;
-int debug;
+int host_trace_level = 1;
```
Comment (lyakh) [inline src/platform/library/lib/trace.c:15]: should be merged with the previous commit to support bisection. Also wouldn't it be better to use `LOG_LEVEL_ERROR`?

File: posix/include/rtos/alloc.h (hunk 1 above)
Comment (lyakh) [reply posix/include/rtos/alloc.h:1]: we do already use SOF headers for library (testbench, cmocka) builds, cannot we also use them for this? Really should avoid duplicating - a sure path to breakage and confusion...

File: src/audio/Kconfig [hunk 4]
```diff
@@ -684,4 +684,11 @@ config WRAP_ACTUAL_POSITION
 	  It is not necessary that on wrap, the buffer position would be zero.At wrap,
 	  in some cases based on the period size, the frame may not exactly be at the
 	  end of the buffer and roll over for some bytes from the beginning of the buffer.
+
+config COMP_MODULES_SO
```
Comment (lyakh) [inline src/audio/Kconfig:688]: we might soon need a similar option to build SOF modules for loading to the DSP, can we maybe add "HOST" somewhere to this option to distinguish between the two?

File: posix/include/rtos/alloc.h (hunk 1 above)
Comment (kv2019i) [reply posix/include/rtos/alloc.h:1]: But which set? I think this approach has merits as well. I think we have three options:
1) XTOS build with CONFIG_LIBRARY (use sof/xtos/rtos)
2) Zephyr build with posix as arch (use sof/zephyr/rtos)
3) Native library build (use sof/posix/rtos)

I think (3) involves code duplication, but for a pure library build (that can be linked to any Linux application), I think this approach is cleanest. Especially if we longterm remove the the XTOS build. Using Zephyr posix build is one option, but I don't think that is really ideal for the library use (like an ALSA plugin, or exposing audio modules via some other framerwork like LV2).

File: posix/include/rtos/atomic.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifndef __SOF_ATOMIC_H__
+#define __SOF_ATOMIC_H__
+
+#include <arch/atomic.h>
+#include <stdint.h>
+
+static inline void atomic_init(atomic_t *a, int32_t value)
+{
+	arch_atomic_init(a, value);
```
Comment (kv2019i) [inline posix/include/rtos/atomic.h:16]: I don't think this arch_foo() layer makes sense. If we pull the rtos layer out from XTOS (or Zephyr), then we don't need any dependency to a separate arch layer. I.e. this should use POSIX primitives (or Linux extensions) directly.

File: src/arch/host/include/arch/debug/panic.h [hunk 6]
```diff
@@ -5,7 +5,7 @@
  * Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
  */
 
-#ifdef __XTOS_RTOS_PANIC_H__
+#ifdef __POSIX_RTOS_PANIC_H__
```
Comment (kv2019i) [inline src/arch/host/include/arch/debug/panic.h:8]: Hmm, continuing with previous, @lrgirdwo @ranj063 , I don't see a point in arch layer if we go for indepdent "rtos" implementation ("XTOS", "Zephyr" and "posix").

File: src/platform/library/include/platform/lib/memory.h [hunk 7]
```diff
@@ -20,6 +20,11 @@
 
 struct sof;
 
+#define EBB_BANKS_IN_SEGMENT		32
+
+#define PLATFORM_LPSRAM_EBB_COUNT	1
+#define PLATFORM_HPSRAM_EBB_COUNT	1
+
```
Comment (kv2019i) [inline src/platform/library/include/platform/lib/memory.h:27]: We really shouldn't need these anymore.

File: src/audio/Kconfig (hunk 4 above)
Comment (kv2019i) [reply src/audio/Kconfig:688]: Or LIBRARY something, that's what this is in the end, right?

**Review by kv2019i — COMMENTED** [review]
Looks pretty good. One bigger issue is the arch layer. If we decide to go with a top-level rtos abstraction (i.e. library is an alternative to a XTOS or Zephyr build), I don't see a point in using the arch layer. We can proceed to merge this version and remove it later, but I'd like to get a consensus before we proceed. Zephyr posix build is of course one option, but not sure whether it works as a plugin library (probably less work to have a native library build for Linux).

File: tools/testbench/common_test.c (hunk 2 above)
Comment (lgirdwood) [reply tools/testbench/common_test.c:274]: ```debug``` colides easily and ```test_bench_trace``` tightly coupled any user code to testbench.

File: src/platform/library/lib/trace.c (hunk 3 above)
Comment (lgirdwood) [reply src/platform/library/lib/trace.c:15]: ack fo bisect.

File: posix/include/rtos/alloc.h (hunk 1 above)
Comment (lgirdwood) [reply posix/include/rtos/alloc.h:1]: I would move list.h and other generic (non RTOS api) level headers to a sof/lib (but this could be another patch).

File: src/arch/host/include/arch/debug/panic.h (hunk 6 above)
Comment (lgirdwood) [reply src/arch/host/include/arch/debug/panic.h:8]: We are stuck a bit today whilst xtos is still being used, i.e. difficult to extract until more xtos related code can be removed (and this depends on progress with Zephyr integration for all platforms).

File: posix/include/rtos/alloc.h (hunk 1 above)
Comment (kv2019i) [reply posix/include/rtos/alloc.h:1]: @lgirdwood Right, but we don't need XTOS arch code if plugin uses its own rtos implementation.  sof/posix/rtos should be independent of sof/src/arch . cmocka uses it, but it was done before we had a rtos abstraction.

**Review by marc-hb — COMMENTED** [review]
I don't think "host" is a good name for what is really just SIMULation. 

tl;dr: "host" = best avoided. Use "simulation" instead.

**Simulation** is the term used by Zephyr, XTensa and the testbench:
- https://docs.zephyrproject.org/latest/boards/posix/index.html
- https://docs.zephyrproject.org/latest/boards/xtensa/xt-sim/doc/index.html
- https://www.cadence.com/en_US/home/tools/ip/tensilica-ip/technologies.html
- https://thesofproject.github.io/latest/developer_guides/testbench/index.html

"Host" is already used for something different: Host is where the Linux / Windows driver run:
- https://thesofproject.github.io/latest/introduction/index.html

Of course "host" is typically the same system for many developers because why would you inconveniently use two different systems when the same workstation can do everything. Yet it absolutely does not have to be the same system because these two "hosts" are conceptually two totally different things. The testbench and the fuzzer have many "driver-free" users. Conversely, production "hosts" use drivers but never any testbench or fuzzer. For Yocto users these two "hosts" are different systems too.

Using the same word "host" for everything just because some people use the same workstation for everything maintains a bad confusion.

Making things even worse, "host" and "target" are also widely misunderstood toolchain concepts:
https://mesonbuild.com/Cross-compilation.html
> In a nutshell, the typical mistake assumes that the terms build, host and target refer to some fixed positions whereas they're actually relative to where the current compiler is running. Think of host as a child of the current compiler and target as an optional grand-child. Compilers don't change their terminology when they're creating another compiler.

**Comment by lgirdwood** [comment]
> I don't think "host" is a good name for what is really just SIMULation.
> 
> tl;dr: "host" = best avoided. Use "simulation" instead.
> 
> **Simulation** is the term used by Zephyr and XTensa:

@marc-hb Not following why you think this is a simulation ? This is an ALSA plugin that can run SOF pipeline topologies and modules in userspace.

**Comment by marc-hb** [comment]
> @marc-hb Not following why you think this is a simulation ? This is an ALSA plugin that can run SOF pipeline topologies and modules in userspace.

Because it does not run on the DSP like all the other simulations I listed above.

File: posix/include/rtos/atomic.h (hunk 5 above)
Comment (ranj063) [reply posix/include/rtos/atomic.h:16]: @kv2019i but the arch layer used here would use the gcc atomic builtins as defined in the host arch atomic.h. Isnt that good enough?

**Comment by ranj063** [comment]
> > @marc-hb Not following why you think this is a simulation ? This is an ALSA plugin that can run SOF pipeline topologies and modules in userspace.
> 
> Because it does not run on the DSP like all the other simulations I listed above.

This is not sumlation @marc-hb and host matches exactly with your quoted definition. These modules can be used to run pipelines on the host just like the linux/windows driver.

**Comment by lgirdwood** [comment]
> > @marc-hb Not following why you think this is a simulation ? This is an ALSA plugin that can run SOF pipeline topologies and modules in userspace.
> 
> Because it does not run on the DSP like all the other simulations I listed above.

@marc-hb You are probably thinking of the POSIX Zephyr mode where SOF runs under Zephyr as a POSIX application (it could be considered a simulation) that we use for fuzzing. This plugin is a userspace executable that runs generic SOF pipeline and module code.

**Comment by marc-hb** [comment]
> @marc-hb You are probably thinking of ...

There are different cases and I'm thinking "host" suits none of them.

In some places in the code we need a name for "not running on DSP". But negations don't make good names; I don't think anyone wants something like CONFIG_NOT_ON_THE_DSP.

So instead we tend to see things like HOST_MODULES, `host_trace_level` and `arch/host/include`. But which host!?! There are _many different hosts_, so "host" is very confusing. 

=> Specific "hosts" should be named with their specific name, e.g.: fuzzer, testbench, zephyr POSIX, ALSA plugin, etc. For Linux/Windows "hosts" it's probably too late and I don't have any better name.

=> CONFIG_NOT_ON_THE_DSP should be called CONFIG_SIMULATION because it's the term already used everywhere, see all the examples above.

**Comment by lyakh** [comment]
> CONFIG_NOT_ON_THE_DSP should be called CONFIG_SIMULATION because it's the term already used everywhere, see all the examples above.

@marc-hb I disagree too. All the above examples are ok with "simulation" because they are simulations. Whereas actually running in a user configuration "on the host," being part of the audio processing pipeline and performing useful work - that isn't a simulation any more. Historically maybe "host" used to mean "native" to the "build host." I.e. not cross-compiled, but now it certainly doesn't have to be the case. E.g. in case of ARM if we build ALSA plugins, we'll build them to run on the ARM "host" and we might well cross-compile them. So in case of ARM we have 3 architectures: build / test / simulate, DSP, and OS "host." Then, yes, we could use "simulation" for the test system, like the CI, which is typically x86*, the DSP can be Xtensa or ARM, and the "OS host" can be x86* or ARM.

File: posix/include/rtos/atomic.h (hunk 5 above)
Comment (kv2019i) [reply posix/include/rtos/atomic.h:16]: @ranj063  This surely works and to be clear, I'm ok to merge this as an intermedia step, but I want to get aligned conceptually that there is no need to depend on the XTOS arch layer.

File: src/include/sof/trace/trace.h [hunk 8]
```diff
@@ -233,19 +233,26 @@ void _log_sofdict(log_func_t sofdict_logf, bool atomic, const void *log_entry,
 
 #ifdef CONFIG_LIBRARY
 
-extern int test_bench_trace;
+#include <sys/time.h>
+
+/* trace level used on host configurations */
+extern int host_trace_level;
```
Comment (kv2019i) [inline src/include/sof/trace/trace.h:239]: I don't think we need to get caught up with the naming (so not a showstopper for me), but as a option, we could call this "extern int sof_library_trace_level".

I think this is logical. We have a build target to build SOF a library, not an application, and when built this way, certain extra interfaces are exposed to do things normally done internally by "SOF-the-application". Like setting trace levels.

File: src/audio/Kconfig [hunk 9]
```diff
@@ -684,4 +684,12 @@ config WRAP_ACTUAL_POSITION
 	  It is not necessary that on wrap, the buffer position would be zero.At wrap,
 	  in some cases based on the period size, the frame may not exactly be at the
 	  end of the buffer and roll over for some bytes from the beginning of the buffer.
+
+config COMP_HOST_MODULES_SO
+	bool "Build modules as shared objects"
```
Comment (kv2019i) [inline src/audio/Kconfig:689]: COMP_MODULE_SHARED_LIBRARY_BUILD

"Build SOF modules as shared libraries"

Comment (ranj063) [reply src/audio/Kconfig]: updated now @kv2019i

**Comment by marc-hb** [comment]
> Whereas actually running in a user configuration "on the host," being part of the audio processing pipeline and performing useful work - that isn't a simulation any more. 

You could call it "hybrid emulation" if you want to make a difference. Note the difference between emulation and simulation depends on whom you ask. Either term makes is clear that the code does not run on the DSP.

"Host" is clearly not on the DSP either but it makes you wonder "which host(s)?" which may or may not be relevant. Super confusing.

> E.g. in case of ARM if we build ALSA plugins, we'll build them to run on the ARM "host" and we might well cross-compile them. So in case of ARM we have 3 architectures: build / test / simulate, DSP, and OS "host." Then, yes, we could use "simulation" for the test system, like the CI, which is typically x86*, the DSP can be Xtensa or ARM, and the "OS host" can be x86* or ARM.

Toolchains matter of course but I'm not sure architecture examples really help clarifying... Anyway good demonstration that the word "host" is now void of any meaning, thanks this is my main point! I suspect "host" is appealing because it's.... short.

**Review by RanderWang — APPROVED** [review]
LGTM

File: src/include/sof/trace/trace.h [hunk 10]
```diff
@@ -233,19 +233,26 @@ void _log_sofdict(log_func_t sofdict_logf, bool atomic, const void *log_entry,
...
-		fprintf(stderr, msg, strrchr(__FILE__, '/') + 1,\
-			__LINE__, ##__VA_ARGS__);	\
-		fprintf(stderr, "\n");							\
-	}										\
+do {								\
+	(void)ctx;						\
+	(void)id_1;						\
+	(void)id_2;						\
+	struct timeval tv;					\
+	char *msg = "(%s:%d) " format;				\
```
Comment (lyakh) [inline src/include/sof/trace/trace.h:248]: Why not put both these variables under the `if`? In practical terms the difference isn't huge - also this way their scope is limited to the "pseudo-loop," but having them under the `if` would be more logical

File: tools/testbench/common_test.c (hunk 2 above)
Comment (lyakh) [reply tools/testbench/common_test.c:274]: @lgirdwood sorry, I mean it's a wrong assignment IMHO. Before `test_bench_trace` was effectively a boolean variable - it only took values of 0 and 1. Now `host_trace_level` is a printing verbosity **level** taking a value between `LOG_LEVEL_CRITICAL` (1) and `LOG_LEVEL_VERBOSE` (4). So this assignment should be fixed.

File: posix/include/rtos/alloc.h (hunk 1 above)
Comment (lyakh) [reply posix/include/rtos/alloc.h:1]: not hugely important, but calling these "posix headers" is a bit confusing - don't think all these headers are POSIX

**Review by lyakh — CHANGES_REQUESTED** [review]

**Review by kv2019i — APPROVED** [review]
I'm good to merge. We probably need to refactor rtos layer more, but this is easier done in a follow-up. Having this merged, allows more people to contribute and this doesn't interfere with other builds of SOF, so I don't see much to gain by keeping this in review for much logner.

**Comment by lyakh** [comment]
> I'm good to merge. We probably need to refactor rtos layer more, but this is easier done in a follow-up. Having this merged, allows more people to contribute and this doesn't interfere with other builds of SOF, so I don't see much to gain by keeping this in review for much logner.

@kv2019i I think https://github.com/thesofproject/sof/pull/8132#discussion_r1316793944 is a bug and should be fixed

**Review by lgirdwood — APPROVED** [review]

File: tools/testbench/common_test.c (hunk 2 above)
Comment (lgirdwood) [reply tools/testbench/common_test.c:274]: @ranj063 can you update testbench to take a log level instead of a bool on the command line which can be passed above. Thanks!

Comment (ranj063) [reply tools/testbench/common_test.c]: fixed now

File: posix/include/sof/lib/mailbox.h [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+static inline
+void mailbox_dspbox_read(void *dest, size_t dest_size,
+			 size_t offset, size_t bytes)
+{
+	int dsp_read_err __unused;
+
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)(MAILBOX_DSPBOX_BASE +
+									offset), bytes);
```
Comment (lyakh) [inline posix/include/sof/lib/mailbox.h:64]: do we really need all these dcache calls? None of the hosts are cache-incoherent, are they?
